File: sdk/python/captcha-font-sdk/captcha_font_sdk/locator.py
from __future__ import annotations

from dataclasses import asdict
from pathlib import Path
import shutil
from typing import Dict, Iterable, Optional

from .background_index import (
    build_background_index as _build_background_index,
    compute_group_id as _compute_group_id,
    load_rgba_pixels,
)
from .component_extractor import extract_components
from .diff_engine import build_diff
from .font_glyph_extractor import CaptchaRgbaImage, build_font_glyphs
from .font_glyph_features import build_font_glyph_features
from .font_glyph_images import export_font_glyph_images
from .font_glyph_slots import align_font_glyph_features_to_slots
from .group_id import group_id_from_pixels
from .text_layer_renderer import render_text_layer
from .text_region_builder import TextRegionFilterConfig, build_text_regions, union_bbox
from .types import (
    BackgroundMeta,
    BackgroundRestoreResult,
    FontGlyphFeatureExtractResult,
    FontGlyphImageExportResult,
    FontGlyphExtractResult,
    FontGlyphSlotExtractResult,
    LocateResult,
    TextLayerResult,
    TextLocateResult,
)


class CaptchaFontLocator:
    """Captcha font locator SDK.

    Flow:
    1) Build mapping: group_id -> background image.
    2) For a captcha image, compute group_id and find matching background.
    3) Diff captcha vs background and extract font components by color-aware flood fill.
    """

    def __init__(
        self,
        diff_threshold: int = 18,
        min_component_pixels: int = 8,
        connectivity: int = 8,
        text_min_width: int = 3,
        text_min_height: int = 3,
        text_min_fill_ratio: float = 0.06,
        text_max_fill_ratio: float = 0.95,
        text_merge_gap: int = 2,
        text_min_vertical_overlap_ratio: float = 0.4,
    ) -> None:
        if connectivity not in (4, 8):
            raise ValueError("connectivity must be 4 or 8")
        self.diff_threshold = diff_threshold
        self.min_component_pixels = min_component_pixels
        self.connectivity = connectivity
        self._text_region_filter = TextRegionFilterConfig(
            min_width=text_min_width,
            min_height=text_min_height,
            min_fill_ratio=text_min_fill_ratio,
            max_fill_ratio=text_max_fill_ratio,
            merge_gap=text_merge_gap,
            min_vertical_overlap_ratio=text_min_vertical_overlap_ratio,
        )
        self._backgrounds: Dict[str, BackgroundMeta] = {}

    @property
    def backgrounds(self) -> Dict[str, BackgroundMeta]:
        return self._backgrounds

    def set_background_index(self, index: Dict[str, BackgroundMeta]) -> None:
        self._backgrounds = dict(index)

    def build_background_index(
        self,
        background_dir: str,
        recursive: bool = True,
        exts: Optional[Iterable[str]] = None,
    ) -> Dict[str, BackgroundMeta]:
        index = _build_background_index(background_dir, recursive=recursive, exts=exts)
        self.set_background_index(index)
        return index

    def compute_group_id(self, image_path: str) -> str:
        return _compute_group_id(image_path)

    def _match_background_meta(
        self,
        captcha_path: str,
    ) -> tuple[str, tuple[int, int], list[tuple[int, ...]], BackgroundMeta]:
        if not self._backgrounds:
            raise RuntimeError("background index is empty, call build_background_index first")

        width, height, cap_px = load_rgba_pixels(captcha_path)
        group_id = group_id_from_pixels(cap_px, width, height)
        if group_id not in self._backgrounds:
            raise KeyError(f"group_id not found in background index: {group_id}")

        background_meta = self._backgrounds[group_id]
        if (background_meta.width, background_meta.height) != (width, height):
            raise ValueError(
                f"background size mismatch: captcha={width}x{height}, background={background_meta.width}x{background_meta.height}"
            )
        return group_id, (width, height), cap_px, background_meta

    def locate_fonts(
        self,
        captcha_path: str,
        include_pixels: bool = True,
    ) -> LocateResult:
        gid, (w, h), cap_px, bg_meta = self._match_background_meta(captcha_path)
        bg_w, bg_h, bg_px = load_rgba_pixels(bg_meta.image_path)
        if (bg_w, bg_h) != (w, h):
            raise ValueError(
                f"background size mismatch: captcha={w}x{h}, background={bg_w}x{bg_h}"
            )

        diff_mask, diff_colors = build_diff(cap_px, bg_px, w, h, self.diff_threshold)
        components = extract_components(
            diff_mask,
            diff_colors,
            w,
            h,
            connectivity=self.connectivity,
            min_component_pixels=self.min_component_pixels,
            include_pixels=include_pixels,
        )

        return LocateResult(
            group_id=gid,
            background_path=bg_meta.image_path,
            image_size=(w, h),
            components=components,
            stats={
                "diff_pixels": sum(1 for v in diff_mask if v),
                "component_count": len(components),
                "min_component_pixels": self.min_component_pixels,
                "diff_threshold": self.diff_threshold,
            },
        )

    def locate_fonts_dict(self, captcha_path: str, include_pixels: bool = True) -> Dict:
        result = self.locate_fonts(captcha_path, include_pixels=include_pixels)
        data = asdict(result)
        data["components"] = sorted(data["components"], key=lambda c: c["pixel_count"], reverse=True)
        return data

    def restore_background(
        self,
        captcha_path: str,
        output_path: Optional[str] = None,
    ) -> BackgroundRestoreResult:
        group_id, image_size, _cap_px, background_meta = self._match_background_meta(captcha_path)
        normalized_output_path = None
        if output_path:
            target = Path(output_path)
            if target.parent and not target.parent.exists():
                target.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(background_meta.image_path, target)
            normalized_output_path = str(target)

        return BackgroundRestoreResult(
            group_id=group_id,
            background_path=background_meta.image_path,
            image_size=image_size,
            output_path=normalized_output_path,
        )

    def restore_background_dict(
        self,
        captcha_path: str,
        output_path: Optional[str] = None,
    ) -> Dict:
        return asdict(self.restore_background(captcha_path, output_path=output_path))

    def restore_background_by_captcha(
        self,
        captcha_path: str,
        output_path: Optional[str] = None,
    ) -> BackgroundRestoreResult:
        return self.restore_background(captcha_path, output_path=output_path)

    def restore_background_by_captcha_dict(
        self,
        captcha_path: str,
        output_path: Optional[str] = None,
    ) -> Dict:
        return self.restore_background_dict(captcha_path, output_path=output_path)

    def locate_text_regions(
        self,
        captcha_path: str,
        include_pixels: bool = True,
    ) -> TextLocateResult:
        font_result = self.locate_fonts(captcha_path, include_pixels=include_pixels)
        regions = build_text_regions(
            font_result.components,
            include_pixels=include_pixels,
            config=self._text_region_filter,
        )
        return TextLocateResult(
            group_id=font_result.group_id,
            background_path=font_result.background_path,
            image_size=font_result.image_size,
            regions=regions,
            stats={
                "component_count": len(font_result.components),
                "region_count": len(regions),
                "text_pixel_count": sum(region.pixel_count for region in regions),
                "diff_pixels": font_result.stats["diff_pixels"],
                "diff_threshold": self.diff_threshold,
            },
        )

    def locate_text_regions_dict(
        self,
        captcha_path: str,
        include_pixels: bool = True,
    ) -> Dict:
        return asdict(self.locate_text_regions(captcha_path, include_pixels=include_pixels))

    def locate_text_positions(
        self,
        captcha_path: str,
        include_pixels: bool = True,
    ) -> TextLocateResult:
        return self.locate_text_regions(captcha_path, include_pixels=include_pixels)

    def locate_text_positions_dict(
        self,
        captcha_path: str,
        include_pixels: bool = True,
    ) -> Dict:
        return self.locate_text_regions_dict(captcha_path, include_pixels=include_pixels)

    def extract_text_layer(
        self,
        captcha_path: str,
        output_path: Optional[str] = None,
        crop_to_content: bool = False,
    ) -> TextLayerResult:
        text_result = self.locate_text_regions(captcha_path, include_pixels=True)
        bbox = union_bbox(text_result.regions)
        rendered = render_text_layer(
            captcha_path=captcha_path,
            regions=text_result.regions,
            text_bbox=bbox,
            output_path=output_path,
            crop_to_content=crop_to_content,
        )
        return TextLayerResult(
            group_id=text_result.group_id,
            background_path=text_result.background_path,
            image_size=text_result.image_size,
            text_bbox=rendered.text_bbox,
            text_pixel_count=rendered.text_pixel_count,
            output_path=rendered.output_path,
            stats={
                "region_count": text_result.stats["region_count"],
                "component_count": text_result.stats["component_count"],
                "diff_pixels": text_result.stats["diff_pixels"],
                "diff_threshold": text_result.stats["diff_threshold"],
            },
        )

    def extract_text_layer_dict(
        self,
        captcha_path: str,
        output_path: Optional[str] = None,
        crop_to_content: bool = False,
    ) -> Dict:
        return asdict(
            self.extract_text_layer(
                captcha_path=captcha_path,
                output_path=output_path,
                crop_to_content=crop_to_content,
            )
        )

    def extract_font_glyphs(
        self,
        captcha_path: str,
        include_pixels: bool = True,
        include_rgba_2d: bool = False,
    ) -> FontGlyphExtractResult:
        locate_result = self.locate_fonts(captcha_path, include_pixels=True)
        captcha_image = None
        if include_rgba_2d:
            width, height, pixels = load_rgba_pixels(captcha_path)
            captcha_image = CaptchaRgbaImage(width=width, height=height, pixels=pixels)

        glyphs = build_font_glyphs(
            locate_result.components,
            captcha_image=captcha_image,
            include_pixels=include_pixels,
            include_rgba_2d=include_rgba_2d,
        )

        return FontGlyphExtractResult(
            group_id=locate_result.group_id,
            background_path=locate_result.background_path,
            image_size=locate_result.image_size,
            glyphs=glyphs,
            stats={
                "glyph_count": len(glyphs),
                "component_count": locate_result.stats["component_count"],
                "diff_pixels": locate_result.stats["diff_pixels"],
                "diff_threshold": self.diff_threshold,
            },
        )

    def extract_font_glyphs_dict(
        self,
        captcha_path: str,
        include_pixels: bool = True,
        include_rgba_2d: bool = False,
    ) -> Dict:
        return asdict(
            self.extract_font_glyphs(
                captcha_path=captcha_path,
                include_pixels=include_pixels,
                include_rgba_2d=include_rgba_2d,
            )
        )

    def extract_font_glyph_features(
        self,
        captcha_path: str,
        target_width: int = 32,
        target_height: int = 32,
        keep_aspect_ratio: bool = True,
    ) -> FontGlyphFeatureExtractResult:
        glyph_result = self.extract_font_glyphs(
            captcha_path=captcha_path,
            include_pixels=False,
            include_rgba_2d=False,
        )
        glyph_features = build_font_glyph_features(
            glyph_result.glyphs,
            target_width=target_width,
            target_height=target_height,
            keep_aspect_ratio=keep_aspect_ratio,
        )
        return FontGlyphFeatureExtractResult(
            group_id=glyph_result.group_id,
            background_path=glyph_result.background_path,
            image_size=glyph_result.image_size,
            target_size=(target_width, target_height),
            glyph_features=glyph_features,
            stats={
                "glyph_count": len(glyph_features),
                "component_count": glyph_result.stats["component_count"],
                "diff_pixels": glyph_result.stats["diff_pixels"],
                "diff_threshold": self.diff_threshold,
            },
        )

    def extract_font_glyph_features_dict(
        self,
        captcha_path: str,
        target_width: int = 32,
        target_height: int = 32,
        keep_aspect_ratio: bool = True,
    ) -> Dict:
        return asdict(
            self.extract_font_glyph_features(
                captcha_path=captcha_path,
                target_width=target_width,
                target_height=target_height,
                keep_aspect_ratio=keep_aspect_ratio,
            )
        )

    def extract_font_glyph_slots(
        self,
        captcha_path: str,
        slot_count: int = 5,
        target_width: int = 32,
        target_height: int = 32,
        keep_aspect_ratio: bool = True,
    ) -> FontGlyphSlotExtractResult:
        feature_result = self.extract_font_glyph_features(
            captcha_path=captcha_path,
            target_width=target_width,
            target_height=target_height,
            keep_aspect_ratio=keep_aspect_ratio,
        )
        vector_length = target_width * target_height
        slots = align_font_glyph_features_to_slots(
            feature_result.glyph_features,
            slot_count=slot_count,
            vector_length=vector_length,
        )
        return FontGlyphSlotExtractResult(
            group_id=feature_result.group_id,
            background_path=feature_result.background_path,
            image_size=feature_result.image_size,
            target_size=feature_result.target_size,
            slot_count=slot_count,
            slots=slots,
            stats={
                "slot_count": slot_count,
                "filled_slots": sum(1 for slot in slots if slot.present),
                "glyph_count": feature_result.stats["glyph_count"],
                "component_count": feature_result.stats["component_count"],
            },
        )

    def extract_font_glyph_slots_dict(
        self,
        captcha_path: str,
        slot_count: int = 5,
        target_width: int = 32,
        target_height: int = 32,
        keep_aspect_ratio: bool = True,
    ) -> Dict:
        return asdict(
            self.extract_font_glyph_slots(
                captcha_path=captcha_path,
                slot_count=slot_count,
                target_width=target_width,
                target_height=target_height,
                keep_aspect_ratio=keep_aspect_ratio,
            )
        )

    def export_font_glyph_images(
        self,
        captcha_path: str,
        output_dir: str,
        file_prefix: Optional[str] = None,
    ) -> FontGlyphImageExportResult:
        glyph_result = self.extract_font_glyphs(
            captcha_path=captcha_path,
            include_pixels=False,
            include_rgba_2d=True,
        )
        prefix = file_prefix or Path(captcha_path).stem
        return export_font_glyph_images(
            group_id=glyph_result.group_id,
            background_path=glyph_result.background_path,
            image_size=glyph_result.image_size,
            glyphs=glyph_result.glyphs,
            output_dir=output_dir,
            file_prefix=prefix,
        )

    def export_font_glyph_images_dict(
        self,
        captcha_path: str,
        output_dir: str,
        file_prefix: Optional[str] = None,
    ) -> Dict:
        return asdict(
            self.export_font_glyph_images(
                captcha_path=captcha_path,
                output_dir=output_dir,
                file_prefix=file_prefix,
            )
        )

File: typescript-sdk/src/locator.ts
import fs from 'fs';
import path from 'path';
import {
  BackgroundMeta,
  BackgroundRestoreResult,
  FontGlyphExtractResult,
  FontGlyphFeatureExtractResult,
  FontGlyphImageExportResult,
  FontGlyphSlotExtractResult,
  GlyphRenderMode,
  GlyphRenderModeLike,
  LocateResult,
  TextLayerResult,
  TextLocateResult,
} from './types';
import { buildBackgroundIndex, computeGroupId, loadRgbaPixelsFromPath } from './background-index';
import { groupIdFromPixels } from './group-id';
import { buildDiff } from './diff-engine';
import { extractComponents } from './component-extractor';
import { buildFontGlyphFeatures } from './font-glyph-features';
import { buildFontGlyphs, buildFontGlyphsFromTextRegions, CaptchaRgbaImage } from './font-glyph-extractor';
import { alignFontGlyphFeaturesToSlots } from './font-glyph-slots';
import { exportFontGlyphImages } from './font-glyph-images';
import { buildTextRegions, TextRegionFilterConfig, unionBBox } from './text-region-builder';
import { renderTextLayer } from './text-layer-renderer';

export class CaptchaFontLocator {
  private diffThreshold: number;
  private minComponentPixels: number;
  private connectivity: number;
  private textRegionFilter: Required<TextRegionFilterConfig>;
  private backgrounds: Record<string, BackgroundMeta> = {};

  constructor(
    diffThreshold: number = 18,
    minComponentPixels: number = 8,
    connectivity: number = 8,
    textMinWidth: number = 3,
    textMinHeight: number = 3,
    textMinFillRatio: number = 0.06,
    textMaxFillRatio: number = 0.95,
    textMergeGap: number = 2,
    textMinVerticalOverlapRatio: number = 0.4,
    textExpectedRegionCount: number | null = 4,
    textForceMergeMaxGap: number = 28
  ) {
    if (![4, 8].includes(connectivity)) {
      throw new Error('connectivity must be 4 or 8');
    }
    this.diffThreshold = diffThreshold;
    this.minComponentPixels = minComponentPixels;
    this.connectivity = connectivity;
    this.textRegionFilter = {
      min_width: textMinWidth,
      min_height: textMinHeight,
      min_fill_ratio: textMinFillRatio,
      max_fill_ratio: textMaxFillRatio,
      merge_gap: textMergeGap,
      min_vertical_overlap_ratio: textMinVerticalOverlapRatio,
      expected_region_count: textExpectedRegionCount,
      force_merge_max_gap: textForceMergeMaxGap,
    };
  }

  get backgroundsIndex() {
    return this.backgrounds;
  }

  setBackgroundIndex(index: Record<string, BackgroundMeta>) {
    this.backgrounds = { ...index };
  }

  async build_background_index(backgroundDir: string, recursive: boolean = true, exts?: Iterable<string>) {
    const index = await buildBackgroundIndex(backgroundDir, recursive, exts);
    this.setBackgroundIndex(index);
    return index;
  }

  async compute_group_id(imagePath: string) {
    return computeGroupId(imagePath);
  }

  private async matchBackgroundMeta(captchaPath: string) {
    if (!Object.keys(this.backgrounds).length) {
      throw new Error('background index is empty, call build_background_index first');
    }
    const { width, height, pixels } = await loadRgbaPixelsFromPath(captchaPath);
    const groupId = groupIdFromPixels(pixels, width, height);
    if (!this.backgrounds[groupId]) {
      throw new Error(`group_id not found in background index: ${groupId}`);
    }
    const backgroundMeta = this.backgrounds[groupId];
    if (backgroundMeta.width !== width || backgroundMeta.height !== height) {
      throw new Error(`background size mismatch: captcha=${width}x${height}, background=${backgroundMeta.width}x${backgroundMeta.height}`);
    }
    return { groupId, imageSize: [width, height] as [number, number], captchaPixels: pixels, backgroundMeta };
  }

  private async buildDiffAndComponents(
    captchaPath: string,
    includePixels: boolean,
    colorSensitive: boolean
  ) {
    const { groupId, imageSize, captchaPixels, backgroundMeta } = await this.matchBackgroundMeta(captchaPath);
    const { width: bgWidth, height: bgHeight, pixels: backgroundPixels } = await loadRgbaPixelsFromPath(backgroundMeta.image_path);
    if (bgWidth !== imageSize[0] || bgHeight !== imageSize[1]) {
      throw new Error(`background size mismatch: captcha=${imageSize[0]}x${imageSize[1]}, background=${bgWidth}x${bgHeight}`);
    }
    const { mask, colors } = buildDiff(captchaPixels, backgroundPixels, imageSize[0], imageSize[1], this.diffThreshold);
    const components = extractComponents(
      mask,
      colors,
      imageSize[0],
      imageSize[1],
      this.connectivity,
      this.minComponentPixels,
      includePixels,
      colorSensitive
    );
    return { groupId, imageSize, backgroundMeta, diffMask: mask, components };
  }

  async locate_fonts(captchaPath: string, includePixels: boolean = true): Promise<LocateResult> {
    const { groupId, imageSize, backgroundMeta, diffMask, components } = await this.buildDiffAndComponents(
      captchaPath,
      includePixels,
      true
    );
    return {
      group_id: groupId,
      background_path: backgroundMeta.image_path,
      image_size: imageSize,
      components,
      stats: {
        diff_pixels: diffMask.filter(Boolean).length,
        component_count: components.length,
        min_component_pixels: this.minComponentPixels,
        diff_threshold: this.diffThreshold,
      },
    };
  }

  async locate_fonts_dict(captchaPath: string, includePixels: boolean = true) {
    return this.locate_fonts(captchaPath, includePixels);
  }

  async restore_background(captchaPath: string, outputPath?: string): Promise<BackgroundRestoreResult> {
    const { groupId, imageSize, backgroundMeta } = await this.matchBackgroundMeta(captchaPath);
    let normalizedOutput: string | null = null;
    if (outputPath) {
      const target = path.resolve(outputPath);
      await fs.promises.mkdir(path.dirname(target), { recursive: true });
      await fs.promises.copyFile(backgroundMeta.image_path, target);
      normalizedOutput = target;
    }
    return {
      group_id: groupId,
      background_path: backgroundMeta.image_path,
      image_size: imageSize,
      output_path: normalizedOutput,
    };
  }

  async restore_background_dict(captchaPath: string, outputPath?: string) {
    return this.restore_background(captchaPath, outputPath);
  }

  async restore_background_by_captcha(captchaPath: string, outputPath?: string) {
    return this.restore_background(captchaPath, outputPath);
  }

  async restore_background_by_captcha_dict(captchaPath: string, outputPath?: string) {
    return this.restore_background(captchaPath, outputPath);
  }

  async locate_text_regions(captchaPath: string, includePixels: boolean = true): Promise<TextLocateResult> {
    const { groupId, imageSize, backgroundMeta, diffMask, components } = await this.buildDiffAndComponents(
      captchaPath,
      includePixels,
      false
    );
    const regions = buildTextRegions(components, includePixels, this.textRegionFilter);
    return {
      group_id: groupId,
      background_path: backgroundMeta.image_path,
      image_size: imageSize,
      regions,
      stats: {
        component_count: components.length,
        region_count: regions.length,
        text_pixel_count: regions.reduce((sum, r) => sum + r.pixel_count, 0),
        diff_pixels: diffMask.filter(Boolean).length,
        diff_threshold: this.diffThreshold,
      },
    };
  }

  async locate_text_regions_dict(captchaPath: string, includePixels: boolean = true) {
    return this.locate_text_regions(captchaPath, includePixels);
  }

  async locate_text_positions(captchaPath: string, includePixels: boolean = true) {
    return this.locate_text_regions(captchaPath, includePixels);
  }

  async locate_text_positions_dict(captchaPath: string, includePixels: boolean = true) {
    return this.locate_text_regions(captchaPath, includePixels);
  }

  async extract_text_layer(captchaPath: string, outputPath?: string, cropToContent: boolean = false): Promise<TextLayerResult> {
    const textResult = await this.locate_text_regions(captchaPath, true);
    const bbox = unionBBox(textResult.regions);
    const rendered = await renderTextLayer(captchaPath, textResult.regions, bbox, outputPath, cropToContent);
    return {
      group_id: textResult.group_id,
      background_path: textResult.background_path,
      image_size: textResult.image_size,
      text_bbox: rendered.text_bbox,
      text_pixel_count: rendered.text_pixel_count,
      output_path: rendered.output_path,
      stats: {
        region_count: textResult.stats.region_count,
        component_count: textResult.stats.component_count,
        diff_pixels: textResult.stats.diff_pixels,
        diff_threshold: textResult.stats.diff_threshold,
      },
    };
  }

  async extract_text_layer_dict(captchaPath: string, outputPath?: string, cropToContent: boolean = false) {
    return this.extract_text_layer(captchaPath, outputPath, cropToContent);
  }

  async extract_font_glyphs(captchaPath: string, includePixels: boolean = true, includeRgba2d: boolean = false): Promise<FontGlyphExtractResult> {
    const locateResult = await this.locate_fonts(captchaPath, true);
    let captchaImage: CaptchaRgbaImage | null = null;
    if (includeRgba2d) {
      const { width, height, pixels } = await loadRgbaPixelsFromPath(captchaPath);
      captchaImage = { width, height, pixels };
    }

    const glyphs = buildFontGlyphs(locateResult.components, captchaImage, includePixels, includeRgba2d);

    return {
      group_id: locateResult.group_id,
      background_path: locateResult.background_path,
      image_size: locateResult.image_size,
      glyphs,
      stats: {
        glyph_count: glyphs.length,
        component_count: locateResult.stats.component_count,
        diff_pixels: locateResult.stats.diff_pixels,
        diff_threshold: this.diffThreshold,
      },
    };
  }

  async extract_font_glyphs_dict(captchaPath: string, includePixels: boolean = true, includeRgba2d: boolean = false) {
    return this.extract_font_glyphs(captchaPath, includePixels, includeRgba2d);
  }

  async extract_font_glyph_features(captchaPath: string, targetWidth: number = 32, targetHeight: number = 32, keepAspectRatio: boolean = true): Promise<FontGlyphFeatureExtractResult> {
    const glyphResult = await this.extract_font_glyphs(captchaPath, false, false);
    const glyphFeatures = buildFontGlyphFeatures(glyphResult.glyphs, targetWidth, targetHeight, keepAspectRatio);
    return {
      group_id: glyphResult.group_id,
      background_path: glyphResult.background_path,
      image_size: glyphResult.image_size,
      target_size: [targetWidth, targetHeight],
      glyph_features: glyphFeatures,
      stats: {
        glyph_count: glyphFeatures.length,
        component_count: glyphResult.stats.component_count,
        diff_pixels: glyphResult.stats.diff_pixels,
        diff_threshold: this.diffThreshold,
      },
    };
  }

  async extract_font_glyph_features_dict(captchaPath: string, targetWidth: number = 32, targetHeight: number = 32, keepAspectRatio: boolean = true) {
    return this.extract_font_glyph_features(captchaPath, targetWidth, targetHeight, keepAspectRatio);
  }

  async extract_font_glyph_slots(captchaPath: string, slotCount: number = 5, targetWidth: number = 32, targetHeight: number = 32, keepAspectRatio: boolean = true): Promise<FontGlyphSlotExtractResult> {
    const featureResult = await this.extract_font_glyph_features(captchaPath, targetWidth, targetHeight, keepAspectRatio);
    const vectorLength = targetWidth * targetHeight;
    const slots = alignFontGlyphFeaturesToSlots(featureResult.glyph_features, slotCount, vectorLength);
    return {
      group_id: featureResult.group_id,
      background_path: featureResult.background_path,
      image_size: featureResult.image_size,
      target_size: featureResult.target_size,
      slot_count: slotCount,
      slots,
      stats: {
        slot_count: slotCount,
        filled_slots: slots.filter((slot) => slot.present).length,
        glyph_count: featureResult.stats.glyph_count,
        component_count: featureResult.stats.component_count,
      },
    };
  }

  async extract_font_glyph_slots_dict(captchaPath: string, slotCount: number = 5, targetWidth: number = 32, targetHeight: number = 32, keepAspectRatio: boolean = true) {
    return this.extract_font_glyph_slots(captchaPath, slotCount, targetWidth, targetHeight, keepAspectRatio);
  }

  async export_font_glyph_images(
    captchaPath: string,
    outputDir: string,
    filePrefix?: string,
    renderMode: GlyphRenderModeLike = GlyphRenderMode.ORIGINAL,
    useTextRegions: boolean = false
  ): Promise<FontGlyphImageExportResult> {
    let glyphs;
    let groupId;
    let backgroundPath;
    let imageSize: [number, number];

    if (useTextRegions) {
      const textResult = await this.locate_text_regions(captchaPath, true);
      const { width, height, pixels } = await loadRgbaPixelsFromPath(captchaPath);
      const captchaImage: CaptchaRgbaImage = { width, height, pixels };
      glyphs = buildFontGlyphsFromTextRegions(textResult.regions, captchaImage, false, true);
      groupId = textResult.group_id;
      backgroundPath = textResult.background_path;
      imageSize = textResult.image_size;
    } else {
      const glyphResult = await this.extract_font_glyphs(captchaPath, false, true);
      glyphs = glyphResult.glyphs;
      groupId = glyphResult.group_id;
      backgroundPath = glyphResult.background_path;
      imageSize = glyphResult.image_size;
    }

    const prefix = filePrefix || path.parse(captchaPath).name;
    return exportFontGlyphImages(groupId, backgroundPath, imageSize, glyphs, outputDir, prefix, renderMode);
  }

  async export_font_glyph_images_dict(
    captchaPath: string,
    outputDir: string,
    filePrefix?: string,
    renderMode: GlyphRenderModeLike = GlyphRenderMode.ORIGINAL,
    useTextRegions: boolean = false
  ) {
    return this.export_font_glyph_images(captchaPath, outputDir, filePrefix, renderMode, useTextRegions);
  }

  async export_text_glyph_images(
    captchaPath: string,
    outputDir: string,
    filePrefix?: string,
    renderMode: GlyphRenderModeLike = GlyphRenderMode.ORIGINAL
  ) {
    return this.export_font_glyph_images(captchaPath, outputDir, filePrefix, renderMode, true);
  }

  async export_text_glyph_images_dict(
    captchaPath: string,
    outputDir: string,
    filePrefix?: string,
    renderMode: GlyphRenderModeLike = GlyphRenderMode.ORIGINAL
  ) {
    return this.export_text_glyph_images(captchaPath, outputDir, filePrefix, renderMode);
  }
}
